### ISSUE TITLE
[S025] Own activation workspaces without losing backward caches

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2#27]
**Prerequisites:** S006, S008, S022–S024; #1 applies. **Timebox:** 40–50 minutes.

## Learn and implement
Create `include/workspace.h`, `src/workspace.c`, and `tests/test_workspace.c`. Allocate the caller-owned arrays required by existing modules for one fixed `(config,B,T)` request: embedding/residual streams; separate per-layer norm, attention, and MLP intermediates; final norm; logits. Use checked counts. Store actual B/T and reject a mismatched forward request. Initially destroy/recreate when shape changes instead of implementing tricky capacity reuse.

Preserve every layer's required forward values until backward; do not reuse a single block cache across all layers. Parameters and workspace have separate owners. Expose a byte-count query without allocation and clean up partial allocation failures. No gradient buffers yet.

## Explicit tests
`tests/test_workspace.c`: enumerate named array shapes and check bounds/nonoverlap for the canonical tiny config. Write per-layer sentinels and verify layers remain independent. Create/free requests with `T=1,3,8` in sequence, reject `T=9`, and force failure at each allocation site using S006's hook. Assert no leaks or published partial workspace. Byte-count calculations must agree with the sum of documented element counts times element sizes.
```sh
make test TEST=workspace
make sanitize TEST=workspace
make test
```
All pass. Deliberately share two layer caches and demonstrate the sentinel test fails.

## Done / teach-back
Explain parameter memory versus activation memory and why training cannot casually overwrite every forward intermediate. Scope is storage and lifecycle, not model execution or memory optimization.
